api/src/Filter/MaterialItemPeriodFilter.php

```diff
@@ -0,0 +1,72 @@
+<?php
+
+namespace App\Filter;
+
+use ApiPlatform\Core\Api\IriConverterInterface;
+use ApiPlatform\Core\Bridge\Doctrine\Orm\Filter\AbstractContextAwareFilter;
+use ApiPlatform\Core\Bridge\Doctrine\Orm\Util\QueryNameGeneratorInterface;
+use App\Entity\Activity;
+use App\Entity\MaterialItem;
+use Doctrine\ORM\Query\Expr\Join;
+use Doctrine\ORM\QueryBuilder;
+use Symfony\Component\PropertyInfo\Type;
+
+final class MaterialItemPeriodFilter extends AbstractContextAwareFilter {
+    public const PERIOD_QUERY_NAME = 'period';
+
+    public function __construct(
+        private IriConverterInterface $iriConverter,
+    ) {
+    }
+
+    // This function is only used to hook in documentation generators (supported by Swagger and Hydra)
+    public function getDescription(string $resourceClass): array {
+        $description = [];
+        $description['period'] = [
+            'property' => self::PERIOD_QUERY_NAME,
+            'type' => Type::BUILTIN_TYPE_STRING,
+            'required' => false,
+        ];
+
+        return $description;
+    }
+
+    protected function filterProperty(string $property, $value, QueryBuilder $q, QueryNameGeneratorInterface $queryNameGenerator, string $resourceClass, string $operationName = null) {
+        if (MaterialItem::class !== $resourceClass) {
+            throw new \Exception("MaterialItemPeriodFilter can only be applies to entities of type MaterialItem (received: {$resourceClass}).");
+        }
+
+        if (self::PERIOD_QUERY_NAME !== $property) {
+            return;
+        }
+
+        // load period from query parameter value
+        $period = $this->iriConverter->getItemfromIri($value);
+
+        // generate alias to avoid interference with other filters
+        $periodParameterName = $queryNameGenerator->generateParameterName($property);
+        $materialNodeJoinAlias = $queryNameGenerator->generateJoinAlias('materialNode');
+        $rootJoinAlias = $queryNameGenerator->generateJoinAlias('root');
+        $ownerJoinAlias = $queryNameGenerator->generateJoinAlias('owner');
+        $activityJoinAlias = $queryNameGenerator->generateJoinAlias('activity');
+        $scheduleEntryJoinAlias = $queryNameGenerator->generateJoinAlias('scheduleEntry');
+
+        $rootAlias = $q->getRootAliases()[0];
+
+        // build relations for materialItems attached via activities
+        $q
+            ->leftJoin("{$rootAlias}.materialNode", $materialNodeJoinAlias)
+            ->leftJoin("{$materialNodeJoinAlias}.root", $rootJoinAlias)
+            ->leftJoin("{$rootJoinAlias}.owner", $ownerJoinAlias)
+            ->leftJoin(Activity::class, $activityJoinAlias, Join::WITH, "{$activityJoinAlias}.id = {$ownerJoinAlias}.id")
+            ->leftJoin("{$activityJoinAlias}.scheduleEntries", $scheduleEntryJoinAlias)
```

Comment:
Dieser Join führt doch dazu, dass wenn eine Activity mehr als einen ScheduleEntry hat, dass dann das MaterialItem mehrfach ausgegeben wird.

Bis und mit Activity ist der Join eindeutig und somit erlaubt.
Für ScheduleEntry braucht es wohl eher eine Extist-Where-Clause

where exists (select 1 from scheduleentry e where e.activity_id = A and e.period_id = P)

Reply:
Good point. Ich hab's kurz getestet und die API gibt keine mehrfachen Werte aus. Ich nehme mal an, dass API platform doppelte Werte ausfiltert. Oder beim Absetzten des SQL geht bereits ein distinct mit.

Reply:
@pmattmann Passt das so für dich, oder möchtest du das geändert sehen?